pkgs/by-name/ty/typescript-language-server/package.nix

```diff
@@ -65,6 +67,8 @@ stdenv.mkDerivation (finalAttrs: {
     makeWrapper "${nodejs}/bin/node" "$out/bin/typescript-language-server" \
       --add-flags "$out/lib/node_modules/typescript-language-server/lib/cli.mjs"
 
+    ${lndir}/bin/lndir ${typescript} $out
```

Comment:
Using lndir here is a bad and hacky solution. Upstream provides the fallbackPath option which is much better.

Reply:
> fallbackPath [string] The path to the tsserver.js file or the typescript lib directory to use when tsserver.path is unspecified/invalid and the tsserver isn't available via the current workspace.

https://github.com/typescript-language-server/typescript-language-server/blob/master/docs/configuration.md

fallbackPath seems like the correct solution which also prefers the project specific version.

Reply:
```patch
diff --git a/pkgs/by-name/ty/typescript-language-server/default-fallbackTsserverPath.diff b/pkgs/by-name/ty/typescript-language-server/default-fallbackTsserv
erPath.diffde 100644
index 000000000000..71cdaf05096a
--- /dev/null
+++ b/pkgs/by-name/ty/typescript-language-server/default-fallbackTsserverPath.diff
@@ -0,0 +1,15 @@
+diff --git a/src/lsp-server.ts b/src/lsp-server.ts
+index ef5907d..9494430 100644
+--- a/src/lsp-server.ts
++++ b/src/lsp-server.ts
+@@ -323,6 +323,10 @@ export class LspServer {
+             }
+         }
+
++        if (!fallbackTsserverPath) {
++            fallbackTsserverPath = "@typescript@";
++        }
++
+         const fallbackVersionProvider = new TypeScriptVersionProvider(fallbackTsserverPath, this.logger);
+         const fallbackSettingVersion = fallbackVersionProvider.getUserSettingVersion();
+         if (fallbackSettingVersion) {
diff --git a/pkgs/by-name/ty/typescript-language-server/package.nix b/pkgs/by-name/ty/typescript-language-server/package.nix
index 053f665ce7eb..b5fb1b92cbae 100644
--- a/pkgs/by-name/ty/typescript-language-server/package.nix
+++ b/pkgs/by-name/ty/typescript-language-server/package.nix
@@ -6,8 +6,10 @@
 , makeWrapper
 , nodejs
 , prefetch-yarn-deps
+, substituteAll
 , yarn
 , testers
+, typescript
 }:

 stdenv.mkDerivation (finalAttrs: {
@@ -21,6 +23,13 @@
     hash = "sha256-FCv0+tA7AuCdGeG6FEiMyRAHcl0WbezhNYLL7xp5FWU=";
   };

+  patches = [
+    (substituteAll {
+      src = ./default-fallbackTsserverPath.diff;
+      inherit typescript;
+    })
+  ];
+
   offlineCache = fetchYarnDeps {
     yarnLock = "${finalAttrs.src}/yarn.lock";
     hash = "sha256-nSMhPfbWD93sGIKehBBE/bh4RzHXFtGAjeyG20m/LWQ=";
```

Reply:
I uhhhh I'm kinda a noob and every time I look at a git patch diff like this I get kinda confused.

I see youre adding a patch file in that, but I dont know where the patch file is. Its not in nixpkgs already I dont think, is it the whole block you put there? Every time patch files enter into things I get lost.

Reply:
He basically directly injected the nix store path of typescript into the source code of the language server because we cannot set the configuration option of the language server in the build process.

Reply:
So, unfortunately, thats all I already understood. I guess my question is, did he include the patch file in this comment?

So is this part what goes in ./default-fallbackTsserverPath.diff ? or is the contents of default-fallbackTsserverPath.diff not included in the message?

```
diff --git a/pkgs/by-name/ty/typescript-language-server/default-fallbackTsserverPath.diff b/pkgs/by-name/ty/typescript-language-server/default-fallbackTsserv
erPath.diffde 100644
index 000000000000..71cdaf05096a
--- /dev/null
+++ b/pkgs/by-name/ty/typescript-language-server/default-fallbackTsserverPath.diff
@@ -0,0 +1,15 @@
+diff --git a/src/lsp-server.ts b/src/lsp-server.ts
+index ef5907d..9494430 100644
+--- a/src/lsp-server.ts
++++ b/src/lsp-server.ts
+@@ -323,6 +323,10 @@ export class LspServer {
+             }
+         }
+
++        if (!fallbackTsserverPath) {
++            fallbackTsserverPath = "@typescript@";
++        }
++
+         const fallbackVersionProvider = new TypeScriptVersionProvider(fallbackTsserverPath, this.logger);
+         const fallbackSettingVersion = fallbackVersionProvider.getUserSettingVersion();
+         if (fallbackSettingVersion) {
```

Reply:
regardless, that does sound like it makes more sense than linking it

So, cool with me. IDK how to add it to my commit tho XD

Reply:
```
diff --git a/src/lsp-server.ts b/src/lsp-server.ts
index ef5907d..9494430 100644
--- a/src/lsp-server.ts
+++ b/src/lsp-server.ts
@@ -323,6 +323,10 @@ export class LspServer {
             }
         }

+        if (!fallbackTsserverPath) {
+            fallbackTsserverPath = "@typescript@";
+        }
+
         const fallbackVersionProvider = new TypeScriptVersionProvider(fallbackTsserverPath, this.logger);
         const fallbackSettingVersion = fallbackVersionProvider.getUserSettingVersion();
         if (fallbackSettingVersion) {
```

Wait, this is the patch file in its final form isnt it.

I think I see whats going on. This is it right?

Reply:
If so I think I finally get what Im looking at XD

Reply:
It builds! testing in a sec :)

Reply:
Yup. Works for me! I'm going to commit and squash this change and you can test it when you can

Reply:
Ohh yeah, now I see what you mean. He put the diff and the changes in nixpkgs then created another diff and pasted that into the comment. So it is a diff of a diff.

Reply:
I figured it out eventually. Im very happy to now understand how this works.

I pushed it, test at your leisure lol but it works for me locally :)

Reply:
and substituteAll just looks for stuff between @@ signs and replaces it im 99% sure so I get that part too. Very cool. Very pleased because this is not the first time ive seen a diff like that and had no idea what was going on, but I get it now, and that makes me happy.